File: 1110-delete-nodes-and-return-forest/1110-delete-nodes-and-return-forest.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int[] toDelete;
    private Queue<TreeNode> queue;
    private TreeNode makeNodes(TreeNode node) {
        if(node == null) return null;
        for(int d : toDelete) { //삭제되는 요소 찾기
            if(node.val == d) {
                if(node.left != null) queue.add(node.left);
                if(node.right != null) queue.add(node.right);
                return null;
            }
        }
        
        TreeNode madeNode = new TreeNode(node.val);
        madeNode.left = makeNodes(node.left);
        madeNode.right = makeNodes(node.right);
        
        return madeNode;
    }
    public List<TreeNode> delNodes(TreeNode root, int[] to_delete) {
        this.queue = new ArrayDeque<>();
        this.toDelete = to_delete;
        
        queue.add(root);
        List<TreeNode> answer = new ArrayList<>();
        
        TreeNode node;
        while(!queue.isEmpty()) {
            root = makeNodes(queue.poll());
            if(root != null) answer.add(root);
        }
        
        return answer;
    }
}

File: 1110-delete-nodes-and-return-forest/1110-delete-nodes-and-return-forest.cpp
/**
 * Definition for a binary tree node->
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    queue<TreeNode*> q;
    vector<int> del;
    TreeNode* makeNodes(TreeNode* node) {
        if(node == NULL) {
            return NULL;
        } else if(find(del.begin(), del.end(), node->val) != del.end()) {
            q.push(node->left);
            q.push(node->right);
            return NULL;
        }
        TreeNode* madeNode = new TreeNode(node->val);
        madeNode->left = makeNodes(node->left);
        madeNode->right = makeNodes(node->right);
        
        return madeNode;
    }
    
    vector<TreeNode*> delNodes(TreeNode* root, vector<int>& to_delete) {
        vector<TreeNode*> answer;
        q.push(root);
        del = to_delete;
        
        while(!q.empty()) {
            if (q.front() == NULL) {
                q.pop();
                continue;
            }
            root = makeNodes(q.front());
            q.pop();
            if(root != NULL) {
                answer.push_back(root);
            }
        }
        
        return answer;
    }
};